### ISSUE TITLE
Status?

### ISSUE THREAD

**Issue by philkunz** [jhermsmeier/node-acme-protocol#2]
Hi,

This is by far the cleanest code for acme implementations in node so far. The rest of the available modules strikes me as kind of messy... I noticed that the npm version is below 1.0.0.

What is the current status on this one? What needs to be done?

I'd be happy to help, but would switch to TypeScript for better in code contracts and code overview than pure JS offers.

regards
Phil

**Comment by jhermsmeier** [comment]
Hey,

> This is by far the cleanest code for acme implementations in node so far. The rest of the available modules strikes me as kind of messy...

I'm humbled, thank you :) And yes, that's the very reason why I started this. I want this to be pretty high quality and understandable, since it'll handle a site's identity – and I don't want to screw that up.

> What is the current status on this one? What needs to be done?

The general idea behind this was to have a lower-level implementation of this protocol, with high fidelity and control, which could be used by higher-level modules to automate the process in a cleaner, more abstract fashion (much like the letsencrypt express modules).

At this time it's only capable of doing everything up to requesting a cert, so the challenge handling and response mechanisms would need implementation, but currently I'm implementing the HTTP link header parsing to properly handle TOS updates and other related resources designated by those. There is also plenty to do in the dependencies – the [JSON Web Algorithms] / [JSON Web Keys] currently only support RSA keys. Then there are the tests, which need an overhaul, but are fairly complex (due to time dependency, nonces and the nature of this protocol). So, there's still a lot that needs to happen, and I'm pretty short on time these days.

> I'd be happy to help, but would switch to TypeScript for better in code contracts and code overview than pure JS offers.

Help would be extremely welcome! I would prefer to stay away from TypeScript though (or anything else that requires transpiling for that matter) – even if there seem to be benefits to it.

**TL;DR:** I'll see to adding everything that needs doing as issues with detailed descriptions on the respective repositories, so you (and others) can see what's going on and start working on those, without having to ask ;)

Thanks for bringing this up!

Cheers,
Jonas

**Comment by philkunz** [comment]
Wonderful! ... Maybe I can still convince you for TypeScript :) TypeScript Intellisense is heaven on Earth and I made a tool that takes all the pain out of configuring, transpiling and testing TypeScript: https://www.npmjs.com/package/npmts

**Comment by jhermsmeier** [comment]
I'm a purist in that regard. The thing with TypeScript, or anything else, is that I just don't want to have the potential complexity to begin with – and I don't think that something like that is necessary to have clean, fast, and well structured code. The amount of tooling required to even run the code & tests without pain doesn't please me either. I just want to have the least amount of moving parts I can get away with – which means; ideally removing dependencies, not adding them.
